app/assets/javascripts/dashboard.js

```diff
@@ -538,26 +557,29 @@ $('body').on('change', 'input[name="roles[worker][]"]', function() {
     MinionPoller.selectedNodes.splice(index, 1);
   }
   toggleMinimumNodesAlert();
+  toggleBootstrapButton();
 });
 
 // when selecting a master
 // selects node and makes it impossible to uncheck
 // enable and keep the previous state of the previous master (selected or not)
 // if user select node as master, it checks it as selected
 $('body').on('change', 'input[name="roles[master][]"]', function() {
-  var $previousMaster = $('input[name="roles[worker][]"]:disabled');
-  var previousMasterValue = parseInt($previousMaster.val(), 10);
-  var checked = MinionPoller.selectedNodes.indexOf(previousMasterValue) !== -1;
-
-  $previousMaster.prop('disabled', false);
-  $previousMaster.prop('checked', checked);
+  var $checkbox = $(this).closest('tr').find('input[name="roles[worker][]"]');
+  var value = parseInt(this.value, 10);
 
   if (this.checked) {
-    var $checkbox = $(this).closest('tr').find('input[name="roles[worker][]"]');
+    MinionPoller.selectedMasters.push(value);
 
     $checkbox.prop('checked', true);
     $checkbox.prop('disabled', true);
+  } else {
+    var index = MinionPoller.selectedMasters.indexOf(value);
+    MinionPoller.selectedMasters.splice(index, 1);
+
+    $checkbox.prop('disabled', false);
```

Comment:
We need to maintain the previous behavior of unchecking the worker/node checkbox if it wasn't previously checked before selecting the node as master. Renaming the variable to `$workerCheckbox` would also help the code understanding.

Scenario 1:

1. No nodes selected yet
1. Select a node as master
1. Unselect as master the node from previous step 
1. Worker checkbox should not be selected

Scenario 2:

1. A node is selected to bootstrap
1. Select a node as master
1. Unselect as master the node from previous step 
1. Worker checkbox should be kept as checked

We can write a feature test for this if it doesn't exist yet for further refactoring.